pkg/server/plugin/nodeattestor/devid/devid.go

```diff
@@ -0,0 +1,487 @@
+		return nil, nil, err
+	}
+
+	// Decode attestation data
+	ekCert, err := x509.ParseCertificate(attData.EKCert)
+	if err != nil {
+		return nil, nil, status.Errorf(codes.InvalidArgument, "cannot parse endorsement certificate: %v", err)
+	}
+
+	devIDPub, err := tpm2.DecodePublic(attData.DevIDPub)
+	if err != nil {
+		return nil, nil, status.Errorf(codes.InvalidArgument, "cannot decode public DevID: %v", err)
+	}
+
+	akPub, err := tpm2.DecodePublic(attData.AKPub)
+	if err != nil {
+		return nil, nil, status.Error(codes.InvalidArgument, "cannot to decode attestation key")
+	}
+
+	ekPub, err := tpm2.DecodePublic(attData.EKPub)
+	if err != nil {
+		return nil, nil, status.Error(codes.InvalidArgument, "cannot decode endorsement key")
+	}
+
+	// Verify the public part of the EK generated from the template is the same
+	// than the one in the EK certificate.
+	err = verifyEKsMatch(ekCert, ekPub)
+	if err != nil {
+		return nil, nil, fmt.Errorf("public key in EK certificate differs from public key created via EK template: %w", err)
+	}
+
+	// Verify EK chain of trust using the provided manufacturer roots.
+	err = verifyEKSignature(ekCert, ekRoots)
+	if err != nil {
+		return nil, nil, status.Errorf(codes.Unauthenticated, "cannot verify EK signature: %v", err)
+	}
+
+	// Verify DevID resides in the same TPM than AK
+	err = verifyDevIDCertification(&akPub, &devIDPub, attData.CertifiedDevID, attData.CertificationSignature)
+	if err != nil {
+		return nil, nil, status.Errorf(codes.Unauthenticated, "cannot to verify that DevID is in the same TPM than AK: %v", err)
+	}
+
+	// Issue a credential activation challenge (to verify AK is in the same TPM than EK)
+	challenge, nonce, err := newCredActivationChallenge(akPub, ekPub)
+	if err != nil {
+		return nil, nil, status.Error(codes.Internal, "cannot generate credential activation challenge")
+	}
+
+	return challenge, nonce, nil
+}
+
+func isDevIDResidencyInfoComplete(attReq *common_devid.AttestationRequest) error {
+	if len(attReq.AKPub) == 0 {
+		return status.Error(codes.InvalidArgument, "missing attestation public key")
+	}
+
+	if len(attReq.DevIDPub) == 0 {
+		return status.Error(codes.InvalidArgument, "missing DevID public key")
+	}
+
+	if len(attReq.EKCert) == 0 {
+		return status.Error(codes.InvalidArgument, "missing endorsement certificate")
+	}
+
+	if len(attReq.EKPub) == 0 {
+		return status.Error(codes.InvalidArgument, "missing endorsement public key")
+	}
+
+	return nil
+}
+
+func verifyEKSignature(ekCert *x509.Certificate, roots *x509.CertPool) error {
+	// Check UnhandledCriticalExtensions for OIDs that we know what to do about
+	// it (e.g. it's safe to ignore)
+	subjectAlternativeNameOID := asn1.ObjectIdentifier{2, 5, 29, 17}
+	if len(ekCert.UnhandledCriticalExtensions) > 0 {
+		unhandledExtensions := []asn1.ObjectIdentifier{}
+		for _, oid := range ekCert.UnhandledCriticalExtensions {
+			if oid.Equal(subjectAlternativeNameOID) {
+				// Subject Alternative Name is not processed at the time.
```

Comment:
I'm not quite sure why we need to do this... if ekCert is a parsed x509.Certificate type, we should have correctly handled SAN OIDs... Also, wouldn't go x509 validation fail if we ran into any unhandled critical extensions?

Reply:
Endorsement certificates contains information about the TPM manufacturer, TPM part number and TPM firmware version in the SAN extension.

For example, this is the SAN extension for one sample certificate provided by the TCG in [TCG Credential Profile EK 2.0, page 34](https://trustedcomputinggroup.org/wp-content/uploads/TCG_IWG_EKCredentialProfile_v2p3_r2_pub.pdf). The same document explains how the SAN is defined for endorsement certificates in section 3.2.9.

```
X509v3 Subject Alternative Name: critical
    DirName:/2.23.133.2.1=id:54434700/2.23.133.2.2=ABCDEF123456/2.23.133.2.3=id:00010023
```

This information is not fully parsed by the golang `x509` package and is stored in the `UnhandledCriticalExtensions` field. This function removes that extension from the list to prevent the server to fail the validation.

Note: Now that you mention it, I noticed that there is a bug in the code that removes all the unhandled critical extensions instead of only the one we expect. I'll fix it.